src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -872,6 +906,38 @@ public void FullGC(IEnumerable<StackValue> gcroots, Executive exe)
                 SingleStep(true);
         }
 
+        // Duplicates the public FullGC method for testing purposes
+        internal void FullGCTest(IEnumerable<StackValue> gcroots, Executive exe, 
+            Dictionary<GCState, (Action, Action)> testNotifications)
+        {
+            if (gcroots == null)
+                throw new ArgumentNullException("gcroots");
+
+            if (exe == null)
+                throw new ArgumentNullException("exe");
+
+            while (gcState != GCState.WaitingForRoots)
+            {
+                (Action start, Action end) fns = (() => { }, () => { });
+                var testCB = testNotifications?.TryGetValue(gcState, out fns) ?? false;
+
+                if (testCB) fns.start();
+                SingleStep(true);
+                if (testCB) fns.end();
+            }
+
+            SetRoots(gcroots, exe);
+            while (gcState != GCState.Pause)
+            {
+                (Action start, Action end) fns = (() => { }, () => { });
+                var testCB = testNotifications?.TryGetValue(gcState, out fns) ?? false;
+
+                if (testCB) fns.start();
+                SingleStep(true);
+                if (testCB) fns.end();
+            }
+        }
```

Comment:
Maybe this can be a good compromise. 
Upsides:
No performance penalty for the live code.
Can run in cicd test suite
Downside
Code duplication and maintainability